ddtrace/contrib/internal/ray/patch.py

```diff
@@ -0,0 +1,439 @@
+            set_maybe_big_tag(task_execute_span, "ray.task.kwargs", kwargs)
+
+            result = wrapped(*args, **kwargs)
+
+            task_execute_span.set_tag_str("ray.task.status", "success")
+            return result
+        except BaseException:
+            task_execute_span.set_tag_str("ray.task.status", "error")
+            raise
+
+
+def traced_submit_task(wrapped, instance, args, kwargs):
+    """Trace task submission, i.e the func.remote() call"""
+    if tracer.current_span() is None:
+        tracer.context_provider.activate(_extract_tracing_context_from_env())
+
+    # Inject dd_trace_ctx args in the function being executed by ray
+    with instance._inject_lock:
+        if instance._function_signature is None:
+            instance._function = _wrap_remote_function_execution(instance._function)
+            instance._function.__signature__ = _inject_dd_trace_ctx_kwarg(instance._function)
+            instance._function_signature = extract_signature(instance._function)
+
+    with tracer.trace(
+        f"{instance._function_name}.remote()", service=os.environ.get(RAY_JOB_NAME), span_type=SpanTypes.RAY
+    ) as span:
+        span.set_tag_str(SPAN_KIND, SpanKind.PRODUCER)
+        _inject_ray_span_tags_and_metrics(span)
+
+        try:
+            set_maybe_big_tag(span, "ray.task.args", kwargs.get("args", {}))
+            set_maybe_big_tag(span, "ray.task.kwargs", kwargs.get("kwargs", {}))
+            _inject_context_in_kwargs(span.context, kwargs)
+
+            resp = wrapped(*args, **kwargs)
+
+            span.set_tag_str("ray.task.submit_status", "success")
+            return resp
+        except BaseException as e:
+            span.set_tag_str("ray.task.submit_status", "error")
+            raise e
+
+
+def traced_submit_job(wrapped, instance, args, kwargs):
+    """Trace job submission. This function is also responsible
+    of creating the root span.
+    It will also inject _RAY_SUBMISSION_ID and _RAY_JOB_NAME
+    in the env variable as some spans will not have access to them
+    trough ray_ctx
+    """
+
+    # Three ways of specifying the job name, in order of precedence:
+    # 1. Metadata JSON: ray job submit --metadata_json '{"job_name": "train.cool.model"}' train.py
+    # 2. Special submission ID format: ray job submit --submission_id "job:train.cool.model,run:38" train.py
+    # 3. Ray entrypoint: ray job submit train_cool_model.py
+    submission_id = kwargs.get("submission_id") or generate_job_id()
+    kwargs["submission_id"] = submission_id
+    entrypoint = kwargs.get("entrypoint", "")
+    job_name = (
+        os.environ.get("DD_SERVICE")
+        or kwargs.get("metadata", {}).get("job_name", "")
+        or get_dd_job_name_from_submission_id(submission_id)
+        or get_dd_job_name_from_entrypoint(entrypoint)
+    )
+
+    # Root span creation
+    job_span = tracer.start_span("ray.job", service=job_name or DEFAULT_JOB_NAME, span_type=SpanTypes.RAY)
```

Comment:
Can we a context manager here 

with tracer.start_span("ray.job", ...):....

That way we don't need to include a try except. We should avoid re-raising exceptions raise by customer code. This modifies the stack trace and makes issues harder to debug (and could result in unnecessary support tickets)

Reply:
I don't think we can, because if the submission goes well, we don't want to stop the span (as it is the root span)